File: src/main/java/com/dongtronic/diabot/data/nutrition/NutritionResponseDTO.java

package com.dongtronic.diabot.data.nutrition;

import com.fasterxml.jackson.annotation.*;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

@JsonInclude(JsonInclude.Include.NON_NULL)
@JsonPropertyOrder({
        "foods"
})
public class NutritionResponseDTO {

    @JsonProperty("foods")
    private List<Food> foods = new ArrayList<Food>();
    @JsonIgnore
    private Map<String, Object> additionalProperties = new HashMap<String, Object>();

    @JsonProperty("foods")
    public List<Food> getFoods() {
        return foods;
    }

    @JsonProperty("foods")
    public void setFoods(List<Food> foods) {
        this.foods = foods;
    }

    @JsonAnyGetter
    public Map<String, Object> getAdditionalProperties() {
        return this.additionalProperties;
    }

    @JsonAnySetter
    public void setAdditionalProperty(String name, Object value) {
        this.additionalProperties.put(name, value);
    }

    public Double getTotalCarbs() {
        return foods.stream().filter(food -> food.getTotalCarbohydrate() != null).mapToDouble(Food::getTotalCarbohydrate).sum();
    }

    public Double getTotalFats() {
        return foods.stream().filter(food -> food.getTotalFat() != null).mapToDouble(Food::getTotalFat).sum();
    }

    public Double getTotalSaturatedFats() {
        return foods.stream().filter(food -> food.getSaturatedFat() != null).mapToDouble(Food::getSaturatedFat).sum();
    }

    public Double getTotalFibers() {
        return foods.stream().filter(food -> food.getDietaryFiber() != null).mapToDouble(Food::getDietaryFiber).sum();
    }

    @Override
    public String toString() {
        StringBuilder returned = new StringBuilder();

        for (Food food : foods) {
            returned.append(food.getFoodName())
                    .append(" (").append(food.getServingQty())
                    .append(" ").append(food.getServingUnit())
                    .append(")\n");
        }

        return returned.toString();
    }
}

File: bot/src/main/java/com/dongtronic/diabot/data/nutrition/NutritionResponseDTO.kt
package com.dongtronic.diabot.data.nutrition

import com.fasterxml.jackson.annotation.*
import java.util.*

@JsonInclude(JsonInclude.Include.NON_NULL)
@JsonPropertyOrder("foods")
class NutritionResponseDTO {
    @get:JsonProperty("foods")
    @set:JsonProperty("foods")
    @JsonProperty("foods")
    var foods: List<Food> = ArrayList()
    @JsonIgnore
    private val additionalProperties: MutableMap<String, Any> = HashMap()

    @JsonAnyGetter
    fun getAdditionalProperties(): Map<String, Any> {
        return additionalProperties
    }

    @JsonAnySetter
    fun setAdditionalProperty(name: String, value: Any) {
        additionalProperties[name] = value
    }

    val totalCarbs: Double
        get() = foods.stream().filter { food: Food -> food.totalCarbohydrate != null }.mapToDouble { food -> food.totalCarbohydrate!!.toDouble() }.sum()

    val totalFats: Double
        get() = foods.stream().filter { food: Food -> food.totalFat != null }.mapToDouble { it.totalFat!!.toDouble() }.sum()

    val totalSaturatedFats: Double
        get() = foods.stream().filter { food: Food -> food.saturatedFat != null }.mapToDouble { it.saturatedFat!!.toDouble() }.sum()

    val totalFibers: Double
        get() = foods.stream().filter { food: Food -> food.dietaryFiber != null }.mapToDouble { it.dietaryFiber!!.toDouble() }.sum()

    override fun toString(): String {
        val returned = StringBuilder()
        for (food in foods) {
            returned.append(food.foodName)
                    .append(" (").append(food.servingQty)
                    .append(" ").append(food.servingUnit)
                    .append(")\n")
        }
        return returned.toString()
    }
}